Support SSL server certificate parameter for MySQL

jeff303 opened the pull request:
Adding ssl-cert config field to MySQL DB details map, to hold the server cert chain in PEM format (similar to what is done in MongoDB driver)

Updating MySQL driver init to map :ssl-cert into :serverSslCert for the JDBC url, when ssl is in use and cert is provided (the MariaDB driver we are using accepts PEM format certificates inline directly for the param value, so no need to shepherd into a temp file)

Update CircleCI config with new be-tests-mysql-ssl-latest-ee job, modeled similarly to the be-tests-oracle-ssl-ee one
 - use extra-env to set all the MySQL DB related vars (for an RDS instance, currently)
 - adding the rds-combined-ca-bundle.pem certificate to resources/certificates
 - loading that cert bundle from resources directory via env var

jeff303 commented:
Fixes #1350.

paoliniluis commented:
It would be a good idea to include the Azure cert chain as well (https://cacerts.digicert.com/DigiCertGlobalRootG2.crt.pem) so users can use postgres and mysql in both public clouds without issues (till we have a self upload option and we can eliminate certificate bundling altogether)

paoliniluis commented:
One more thing: if you bundle the scripts in the release jar, they won't be needed in the container right? (so this one can be reverted https://github.com/metabase/metabase/pull/15194/commits/e2147a2647a37cdd5452b5e414975a422186ea1c)

jeff303 commented:
> It would be a good idea to include the Azure cert chain as well (https://cacerts.digicert.com/DigiCertGlobalRootG2.crt.pem) so users can use postgres and mysql in both public clouds without issues (till we have a self upload option and we can eliminate certificate bundling altogether)

> One more thing: if you bundle the scripts in the release jar, they won't be needed in the container right? (so this one can be reverted e2147a2)

I think creating a Metabase-wide trust store and automatically including certain CAs into it is out of scope for this issue/change.  For one thing, it's complicated (we would either need to copy the JDK-wide `cacerts` file, add some stuff to it, then bring it back), and we would either have to override JDK level params (which might have unintended side effects on other stuff), or sneakily create the trust store "behind the scenes" and set it for MySQL, but only in certain conditions (ex: not if the user has their own custom certificate, like for an on-prem database).  That trickiness will become too difficult to manage, IMHO (we would have to present some type of workflow/wizard, first saying "Are you using RDS?", etc. to show/hide various other configs).

I think bundling them with the Docker image is good, since it will automatically handle a lot of this for people running the Docker image (which includes a lot of deployment scenarios).

jeff303 commented:
To close the loop on a few things, as this is finally ready for review.

1. The first problem with the CI tests was that I was using the `>` operator in the CircleCI YAML to try to prepend the extra environment options.  This includes a newline at the end of the string, and hence prepending them wasn't actually working at all.  Fixed by switching to `>-` which chomps the final newline.
2. Snuck in a fix for a [ridiculously obscure problem](https://stackoverflow.com/questions/64122437/java-sql-sqlrecoverableexception-io-error-connection-reset-by-peer#comment113391289_64122437) that was making the Oracle SSL job fail
3. I created a new MySQL RDS instance (tiny) for the express purpose of testing SSL connectivity.  It lives in `us-east-2`, and is called `mysql-ci-test-db`.  The host name, and password for the DB user (whose name is `metabase`) are configured via the CircleCI job environment variables admin page, and the open source CircleCI config here is just referencing them when it sets the relevant `MB_MYSQL_TEST_*` variables for `lein test`

camsaul commented on .circleci/config.yml:
```diff
@@ -1034,6 +1034,22 @@ workflows:
             name: mysql-latest
           driver: mysql
 
+      - test-driver:
+          name: be-tests-mysql-ssl-latest-ee
+          description: "(MySQL latest - connecting with SSL)"
+          requires:
+            - be-tests-ee
+          driver: mysql
+          test-selector: :only metabase.driver.mysql-test
+          extra-env: >-
+            MB_DB_TYPE=h2
+            MB_MYSQL_TEST_HOST=$MYSQL_RDS_SSL_INSTANCE_HOST
+            MB_MYSQL_TEST_SSL=true
+            MB_MYSQL_TEST_ADDITIONAL_OPTIONS='verifyServerCertificate=true'
+            MB_MYSQL_TEST_SSL_CERT="$(cat /home/circleci/metabase/metabase/resources/certificates/rds-combined-ca-bundle.pem)"
+            MB_MYSQL_TEST_USER=metabase
+            MB_MYSQL_TEST_PASSWORD=$MYSQL_RDS_SSL_INSTANCE_PASSWORD
+
```
I think I mentioned this before, but I don't think we do a whole separate test job just for SSL. Really the only thing that's different between the two is whether we can connect or not... Can't you just pass all the connection details to one of the existing MySQL test jobs and have the connection-related tests test both the SSL and non-SSL versions?

Also, this isn't really "latest" the same way the "latest" Docker images are. It might have been "latest" whenever you spun up the RDS instance, but it probably won't be in a year.

jeff303 replied:
Something like this?  (See `mysql-connect-with-ssl-and-pem-cert-test`, which has now been added).

If this is the approach you had in mind, I will split out all of the Oracle related fixes from this PR, and move them to a separate one, and do the same thing there (adding a different test that overrides the respective Oracle env vars to test SSL connectivity in the same test).

jeff303 replied:
I took the liberty of doing that in: #15260.

camsaul reviewed (changes requested):
It seems really crazy to do a whole separate job just to test connecting with SSL. I'd just pass the existing env vars to an existing test job and then write a new test that will look for the extra env vars and see whether we can connect with them if those env vars are present

jeff303 committed:
Support SSL server certificate parameter for MySQL

Adding ssl-cert config field to MySQL DB details map, to hold the server cert chain in PEM format (similar to what is done in MongoDB driver)

Updating MySQL driver init to map :ssl-cert into :serverSslCert for the JDBC url, when ssl is in use and cert is provided (the MariaDB driver we are using accepts PEM format certificates inline directly for the param value, so no need to shepherd into a temp file)

Adding new test to mysql_test.clj to run a single test while connecting via SSL with PEM cert

Update CircleCI config:
 - use extra-env to set all the MySQL SSL instance DB related vars (for an RDS instance, currently)
 - adding the rds-combined-ca-bundle.pem certificate to resources/certificates
 - loading that cert bundle from resources directory via env var

Adding to/fixing assertion in connection-spec-test for :ssl

camsaul reviewed (approved):
cool works for me
